### PR TITLE
Flag Dq for wavelength outside range

### PR DESCRIPTION
**Pull request by jemorrison**

### CODE REVIEW

**Comment by jemorrison** [comment]
I updated the DQ flags for pixels outside the wavelength range of the master background.
I pigged backed off of subtract_2d_background - I believe this should handle all cases.
I only updated the DQ flag with DO_NOT_USE because there is not a NO_BACKGROUND - or something else I can use. 
I am not sure what do with the conflicts that I have with this PR ?

File: jwst/master_background/master_background_step.py [hunk 1]
```diff
@@ -55,7 +55,7 @@ def process(self, input):
 
         with datamodels.open(input) as input_data:
             background_data = None
-
+
```
Comment (philhodge) [inline jwst/master_background/master_background_step.py]: Please delete trailing blanks.

File: jwst/master_background/master_background_step.py [hunk 2]
```diff
@@ -152,7 +152,7 @@ def process(self, input):
             else:
                 background_2d = expand_to_2d(input_data, self.user_background)
                 result = subtract_2d_background(input_data, background_2d)
-
+
```
Comment (philhodge) [inline jwst/master_background/master_background_step.py]: Please delete trailing blanks.

**Comment by jemorrison** [comment]
@jdavies-st could you check out the continuous-integration/jenkins failure - I do not understand what it is trying to tell me - 
Something about an Keyword 'IMX' not found ?

**Comment by jdavies-st** [comment]
The Jenkins CI failure is unrelated.  You can get rid of it by rebasing to `jwst/master` and force push. Or just ignore it.

File: jwst/master_background/expand_to_2d.py [hunk 3]
```diff
@@ -166,13 +170,15 @@ def bkg_for_multislit(input, tab_wavelength, tab_background):
...
+        # flag values outside of background wavelength table
+        mask_limit = (wl_array > max_wave) | (wl_array < min_wave)
+        wl_array[mask_limit] = -1
         # bkg_flux will be a 2-D array, because wl_array is 2-D.
         bkg_flux = np.interp(wl_array, tab_wavelength, tab_background,
                              left=0., right=0.)
 
         background.slits[k].data[:] = bkg_flux.copy()
-
+        background.slits[k].dq[mask_limit] |= dqflags.pixel['DO_NOT_USE']
```
Comment (hbushouse) [inline jwst/master_background/expand_to_2d.py]: Does the ``|=`` operation here work the same as ``numpy.bitwise_or()``? We really do want a bitwise operation, not an operation based on the "total" value of each array element.

File: jwst/master_background/expand_to_2d.py [hunk 4]
```diff
@@ -199,20 +205,23 @@ def bkg_for_image(input, tab_wavelength, tab_background):
...
+    # flag values outside of background wavelength table
+    mask_limit = (wl_array > max_wave) | (wl_array < min_wave)
+    wl_array[mask_limit] = -1
     # bkg_flux will be a 2-D array, because wl_array is 2-D.
     bkg_flux = np.interp(wl_array, tab_wavelength, tab_background,
                          left=0., right=0.)
 
     background.data[:] = bkg_flux.copy()
-
+    background.dq[mask_limit] |= dqflags.pixel['DO_NOT_USE']
```
Comment (hbushouse) [inline jwst/master_background/expand_to_2d.py]: Same question here regarding bitwise_or

File: jwst/master_background/expand_to_2d.py [hunk 5]
```diff
@@ -235,19 +244,29 @@ def bkg_for_ifu_image(input, tab_wavelength, tab_background):
...
             x, y = grid_from_bounding_box(ifu_wcs.bounding_box)
             wl_array = ifu_wcs(x, y)[2]
-
             wl_array[np.isnan(wl_array)] = -1.
+            # flag values outside of the background wavelength table
+            mask_limit = (wl_array > max_wave) | (wl_array < min_wave)
+            wl_array[mask_limit] = -1
+
+            # TODO - add another DQ Flag something like NO_BACKGROUND when we have space in dqflags
+            background.dq[mask_limit] |= dqflags.pixel['DO_NOT_USE']
```
Comment (hbushouse) [inline jwst/master_background/expand_to_2d.py]: And again a question here regarding bitwise_or

File: jwst/master_background/expand_to_2d.py [hunk 6]
```diff
@@ -256,12 +275,18 @@ def bkg_for_ifu_image(input, tab_wavelength, tab_background):
...
-        wl_array[np.isnan(wl_array)] = -1.
+        # first remove the nans from wl_array and replace with -1
+        mask = np.isnan(wl_array)
+        wl_array[mask] = -1.
+        # next look at the limits of the wavelength table
+        mask_limit = (wl_array > max_wave) | (wl_array < min_wave)
+        wl_array[mask_limit] = -1
+
+        # TODO - add another DQ Flag something like NO_BACKGROUND when we have space in dqflags
+        background.dq[mask_limit] |= dqflags.pixel['DO_NOT_USE']
```
Comment (hbushouse) [inline jwst/master_background/expand_to_2d.py]: And here

File: jwst/master_background/master_background_step.py [hunk 7]
```diff
@@ -256,11 +262,12 @@ def _subtract_2d_background(model, background):
         if isinstance(model, datamodels.MultiSlitModel):
             for slit, slitbg in zip(result.slits, background.slits):
                 slit.data -= slitbg.data
+                slit.dq |= slitbg.dq
 
         # Handle MIRI LRS, MIRI MRS and NIRSpec IFU
         elif isinstance(model, (datamodels.ImageModel, datamodels.IFUImageModel)):
             result.data -= background.data
-
+            result.dq |= background.dq
```
Comment (hbushouse) [inline jwst/master_background/master_background_step.py]: One more time: does this give a correct bitwise-or result?

File: jwst/master_background/expand_to_2d.py (hunk 3 above)
Comment (jdavies-st) [reply jwst/master_background/expand_to_2d.py]:
```
In [2]: import numpy as np

In [8]: a = np.array((0, 1, 2, 4), dtype=np.uint32)

In [9]: b = np.array((0, 0, 1, 1), dtype=np.uint32)

In [10]: a | b
Out[10]: array([0, 1, 3, 5], dtype=uint32)
```

Comment (jdavies-st) [reply jwst/master_background/expand_to_2d.py]: https://docs.scipy.org/doc/numpy/reference/generated/numpy.bitwise_or.html

> Computes the bit-wise OR of the underlying binary representation of the integers in the input arrays. This ufunc implements the C/Python operator `|`.

Comment (hbushouse) [reply jwst/master_background/expand_to_2d.py]: And even more robustly:
`In [5]: a=np.array((0,1,2,4,3,6,7),dtype=np.uint32)

In [6]: b=np.array((0,0,0,1,1,1,1),dtype=np.uint32)

In [7]: a | b
Out[7]: array([0, 1, 2, 5, 3, 7, 7], dtype=uint32)
`
shows that for elements that already have a 1 set, it does not add another 1. So it looks good.

Comment (hbushouse) [reply jwst/master_background/expand_to_2d.py]: But I also like the idea of using the explicit ``np.bitwise_or``, just so it's obvious.

Comment (jdavies-st) [reply jwst/master_background/expand_to_2d.py]: The only advantage of `np.bitwise_or()` is that if you need to do use some of the `kwargs` that are standard for `numpy` ufuncs, i.e. `axis=`, etc.  So `np.bitwise_or()` adds some extra functionality on top of `|`, but that's it.

Comment (jdavies-st) [reply jwst/master_background/expand_to_2d.py]: And
```
In [8]: a = np.array((0, 1, 2, 4), dtype=np.uint32)

In [32]: a | 1
Out[32]: array([1, 1, 3, 5], dtype=uint32)
```

File: jwst/master_background/master_background_step.py [hunk 8]
```diff
@@ -52,6 +52,10 @@ def process(self, input):
         """
 
         with datamodels.open(input) as input_data:
+            background_data = None
+            # Handle individual NIRSpec FS, NIRSpec MOS
+            if isinstance(input_data, datamodels.MultiSlitModel):
+                pass
```
Comment (jdavies-st) [inline jwst/master_background/master_background_step.py]: This block should not be here.

File: jwst/master_background/master_background_step.py [hunk 9]
```diff
@@ -114,15 +118,18 @@ def process(self, input):
...
                     result.update(input_data)
                     for model in input_data:
                         background_2d = expand_to_2d(model, master_background)
                         result.append(subtract_2d_background(model, background_2d))
 
-                # Skip step for case with no container and no user background
+                # Record name of user-supplied master background spectrum
+                if isinstance(result, datamodels.ModelContainer):
+                    for model in result:
+                        model.meta.background.master_background_file = basename(self.user_background)
```
Comment (jdavies-st) [inline jwst/master_background/master_background_step.py]: I don't think this block should be here?

Comment (jemorrison) [reply jwst/master_background/master_background_step.py]: I messed up the rebase and conflicts. Dang it . Thanks for catching it.

**Comment by jemorrison** [comment]
@jdavies-st I made the changes people requested. I still get a travis failure but I think that is related
to something else. I think we can now merge.

**Comment by jdavies-st** [comment]
Travis failures are DEV dependency failures due to a change to the ASDF schema tester pytest plugin merged today in the `asdf` repo.  Unrelated, so OK to merge.  👍

**Pull request merged**